code/headers/hardware_usart.hpp

```diff
@@ -5,151 +5,254 @@
     };
 
-    enum class peripheral {
-        peripheral_a,
-        peripheral_b
+    /**
+     * @brief Struct with the data for usart1.
+     *
+     * Usart1 tx is pin: D16.
+     * Usart1 rx is pin: D17.
+     *
+     * With labels TX2 and RX2 on the arduino due.
+     *
+     */
+    struct usart1 {
+        constexpr static uint32_t rx = PIO_PA12A_RXD1;
+        constexpr static uint32_t tx = PIO_PA13A_TXD1;
+
+        using periph = detail::pio::periph_a;
+        using pio = detail::pio::pioa;
     };
 
-    struct hw_usart_s {
-        Usart *usart;
-        uint32_t rx_mask;
-        uint32_t tx_mask;
-        Pio *pio;
-        peripheral periph;
-        uint16_t id;
+    /**
+     * @brief Struct with the data for usart2.
+     *
+     * Usart2 tx is pin: A11.
+     * Usart2 rx is pin: D52.
+     *
+     * This usart port doesn't have TX/RX labels for the pins. 
+     * Just A11 and 52.
+     *
+     */
+    struct usart2 {
+        constexpr static uint32_t rx = PIO_PB21A_RXD2;
+        constexpr static uint32_t tx = PIO_PB20A_TXD2;
+
+        using periph = detail::pio::periph_a;
+        using pio = detail::pio::piob;
     };
 
-    const static hw_usart_s usart[uint8_t(usart_ports::UART_SIZE)] = {
-            {USART0, PIO_PA10A_RXD0, PIO_PA11A_TXD0, PIOA, peripheral::peripheral_a, ID_USART0},
-            {USART1, PIO_PA12A_RXD1, PIO_PA13A_TXD1, PIOA, peripheral::peripheral_a, ID_USART1},
-            {USART3, PIO_PD5B_RXD3, PIO_PD4B_TXD3, PIOD, peripheral::peripheral_b, ID_USART3}
+    /**
+     * @brief Struct with the data for usart3.
+     *
+     * Usart3 tx is pin: D14.
+     * Usart3 rx is pin: D15.
+     *
+     * With labels TX3 and RX3 on the arduino due.
+     *
+     */
+    struct usart3 {
+        constexpr static uint32_t rx = PIO_PD5B_RXD3;
+        constexpr static uint32_t tx = PIO_PD4B_TXD3;
+
+        using periph = detail::pio::periph_b;
+        using pio = detail::pio::piod;
     };
 
-    template <size_t BufferLength = 256>
+    namespace detail {
+        namespace pio {
+            // helper structs for getting the pointer to the pio
+            template <typename PIO>
+            Pio *const port = nullptr;
+
+            template <>
+            Pio *const port<pioa> = PIOA;
+
+            template <>
+            Pio *const port<piob> = PIOB;
+
+            template <>
+            Pio *const port<piod> = PIOD;
+        } // namespace pio
+
+        namespace usart {
+            // helper structs for getting the pointer to the usart
+            template <typename USART>
+            Usart *const port = nullptr;
+
+            template <>
+            Usart *const port<usart0> = USART0;
+
+            template <>
+            Usart *const port<usart1> = USART1;
+
+            template <>
+            Usart *const port<usart2> = USART2;
+
+            template <>
+            Usart *const port<usart3> = USART3;
+        } // namespace usart
+
+        namespace periph_id {
```

Comment:
Nitpick on naming: I'd prefer to see either a full abbreviation or the full word instead of this half-way between the two, unsure if this is something to be enforced, but is worth thinking about in my opinion